File: src/engine/ai/policy/features.ts
/**
 * TIKITAQ — TS-Feature-Encoder für ONNX-Inferenz.
 *
 * Exakter 1:1-Spiegel von `ml/features.py`. Muss IMMER synchron gehalten
 * werden — wenn sich eine Feature-Dimension ändert, muss sie auf beiden
 * Seiten geändert werden, sonst liefert das Netz Unsinn.
 *
 * Die Produktions-Nutzung: von ONNX-Policy in der TS-Arena (onnxPolicy.ts)
 * und später für Browser-Inferenz (onnxruntime-web).
 */

import type { GameState, PlayerData, TeamSide } from '../../types'
import type { BallOption } from '../playerDecision/types'
import type { MatchIntent } from '../matchIntent'

// ── Konstanten (spiegeln Python) ──────────────────────────────

export const ROLE_LABELS = ['TW', 'IV', 'LV', 'RV', 'ZDM', 'LM', 'RM', 'OM', 'ST'] as const
export const OPTION_TYPES = [
  'shoot', 'short_pass', 'long_ball', 'through_ball',
  'cross', 'dribble', 'advance', 'hold',
] as const
export const INTENT_SIDES = ['left', 'center', 'right'] as const

export const PLAYER_FEAT_DIM = 13
export const GLOBAL_FEATURE_DIM = 3 + PLAYER_FEAT_DIM + 4 + 3 + 10 * PLAYER_FEAT_DIM + 11 * PLAYER_FEAT_DIM
export const OPTION_FEATURE_DIM = 8 + 2 + 1 + 1 + 1 + 2

// ── Einzel-Encoder ──────────────────────────────────────────

function roleOneHot(label: string): number[] {
  const out = new Array<number>(ROLE_LABELS.length).fill(0)
  const idx = (ROLE_LABELS as readonly string[]).indexOf(label)
  if (idx >= 0) out[idx] = 1
  return out
}

function optionTypeOneHot(t: string): number[] {
  const out = new Array<number>(OPTION_TYPES.length).fill(0)
  const idx = (OPTION_TYPES as readonly string[]).indexOf(t)
  if (idx >= 0) out[idx] = 1
  return out
}

function intentOneHot(side: 'left' | 'center' | 'right' | null): number[] {
  const out = new Array<number>(INTENT_SIDES.length).fill(0)
  const target = side ?? 'center'
  const idx = (INTENT_SIDES as readonly string[]).indexOf(target)
  if (idx >= 0) out[idx] = 1
  return out
}

function encodePlayer(p: PlayerData): number[] {
  return [
    p.position.x / 100,
    p.position.y / 100,
    ...roleOneHot(p.positionLabel),
    p.fitness / 100,
    p.confidence / 100,
  ]
}

function ghostPlayer(): number[] {
  return new Array<number>(PLAYER_FEAT_DIM).fill(0)
}

function encodeOption(opt: BallOption): number[] {
  const hasReceiver = opt.receiverId != null
  return [
    ...optionTypeOneHot(opt.type),
    opt.target.x / 100,
    opt.target.y / 100,
    opt.successChance,
    opt.reward,
    hasReceiver ? 1 : 0,
    hasReceiver ? opt.target.x / 100 : 0,
    hasReceiver ? opt.target.y / 100 : 0,
  ]
}

// ── Haupt-Encoder ───────────────────────────────────────────

export interface EncodedSample {
  /** Float32Array der Länge GLOBAL_FEATURE_DIM */
  globalFeat: Float32Array
  /** Float32Array der Länge maxOptions × OPTION_FEATURE_DIM */
  optionsFlat: Float32Array
  /** Float32Array der Länge maxOptions — 1 für valide, 0 für Padding */
  mask: Float32Array
  /** Wie viele valide Optionen existieren (< maxOptions) */
  numOptions: number
}

/**
 * Encodet einen GameState + Optionen in Tensor-fähige Float32Arrays,
 * bereit für onnxruntime-node Inferenz.
 */
export function encodeStateForPolicy(
  state: GameState,
  team: TeamSide,
  carrier: PlayerData,
  options: BallOption[],
  intent: MatchIntent | null,
  maxOptions: number = 16,
): EncodedSample {
  const ball = state.ball.position
  const carrierFeats = encodePlayer(carrier)

  const score = state.score
  const ownScore = team === 1 ? score.team1 : score.team2
  const oppScore = team === 1 ? score.team2 : score.team1
  const scoreDiff = (ownScore - oppScore) / 10

  const intentSide = intent?.attackSide ?? null
  const turnIdx = Math.floor(state.gameTime * 2)
  const intentTurnsValid = intent
    ? Math.max(0, intent.validUntilTurn - turnIdx) / 5
    : 0

  // Teammates & Opponents (ohne carrier)
  const allTeammates = state.players.filter(p => p.team === team && p.id !== carrier.id)
  const allOpponents = state.players.filter(p => p.team !== team)

  const teamFeats: number[] = []
  for (let i = 0; i < 10; i++) {
    teamFeats.push(...(i < allTeammates.length ? encodePlayer(allTeammates[i]) : ghostPlayer()))
  }

  const oppFeats: number[] = []
  for (let i = 0; i < 11; i++) {
    oppFeats.push(...(i < allOpponents.length ? encodePlayer(allOpponents[i]) : ghostPlayer()))
  }

  const globalVec = [
    ball.x / 100,
    ball.y / 100,
    1.0,  // possession indicator
    ...carrierFeats,
    scoreDiff,
    state.gameTime / 90,
    intentTurnsValid,
    team,
    ...intentOneHot(intentSide),
    ...teamFeats,
    ...oppFeats,
  ]

  if (globalVec.length !== GLOBAL_FEATURE_DIM) {
    throw new Error(
      `global dim mismatch: expected ${GLOBAL_FEATURE_DIM}, got ${globalVec.length}`,
    )
  }

  const globalFeat = new Float32Array(globalVec)

  // Options-Matrix (flach: maxOptions × OPTION_FEATURE_DIM)
  const optionsFlat = new Float32Array(maxOptions * OPTION_FEATURE_DIM)
  const mask = new Float32Array(maxOptions)
  const numOptions = Math.min(options.length, maxOptions)

  for (let i = 0; i < numOptions; i++) {
    const optFeats = encodeOption(options[i])
    for (let j = 0; j < OPTION_FEATURE_DIM; j++) {
      optionsFlat[i * OPTION_FEATURE_DIM + j] = optFeats[j]
    }
    mask[i] = 1
  }

  return { globalFeat, optionsFlat, mask, numOptions }
}

File: ml/features.py
"""
TIKITAQ ML — Feature-Encoder.

Konvertiert einen JSONL-Record (Ballführer-Entscheidung) in Tensoren:
- `global`: Kontextfeatures (Ball, Carrier, Intent, alle 22 Spieler-
  Positionen und Rollen)
- `options`: pro Option ein Feature-Vektor (Type, Target, Success,
  Reward, Empfänger-Info)
- `mask`: 1 wenn Option gültig, 0 wenn Padding (wichtig für Softmax)
- `chosen`: Index der gewählten Option

Design-Entscheidungen (Version 1, pragmatisch):
- Flache Vektoren statt Grid-CNN — einfacher, schnell genug
- Spieler werden nach Team × Rollen-Slot sortiert: fester Index 0-10
  für eigenes Team, 11-21 für Gegner. Feature-Engineering-light.
- Positionen werden normalisiert (x, y / 100 → [0, 1])
- Stats auch normalisiert (/ 100 → [0, 1])
- Intent als one-hot

Wenn BC schwach performt, hier verfeinern: Grid-Encoding, Player-
Attention, etc.
"""

from __future__ import annotations

from typing import Any

import torch

# ── Feld-Konstanten ─────────────────────────────────────────────

ROLE_LABELS = ["TW", "IV", "LV", "RV", "ZDM", "LM", "RM", "OM", "ST"]
OPTION_TYPES = [
    "shoot", "short_pass", "long_ball", "through_ball",
    "cross", "dribble", "advance", "hold",
]
INTENT_SIDES = ["left", "center", "right"]

# Player-Feature pro Spieler: 2 pos + 9 role one-hot + fitness + confidence = 13
PLAYER_FEAT_DIM = 13

# Global-Feature-Dimension:
#   3 (ball xy + possession-indicator=1)
#   + 13 (carrier features)
#   + 4 (score_diff, game_time, intent_turns_valid, team)
#   + 3 (intent one-hot)
#   + 10 * 13 (teammates; max 10, Position-Label-basiert einsortiert)
#   + 11 * 13 (opponents; max 11)
GLOBAL_FEATURE_DIM = 3 + PLAYER_FEAT_DIM + 4 + 3 + 10 * PLAYER_FEAT_DIM + 11 * PLAYER_FEAT_DIM

# Option-Feature-Dimension:
#   8 (type one-hot)
#   + 2 (target xy)
#   + 1 (success_chance)
#   + 1 (reward)
#   + 1 (has_receiver flag)
#   + 2 (receiver pos — 0 wenn kein Empfänger)
OPTION_FEATURE_DIM = 8 + 2 + 1 + 1 + 1 + 2


# ── Einzel-Encoder ──────────────────────────────────────────────

def _role_onehot(label: str) -> list[float]:
    """9-dimensional one-hot Vektor für positionLabel."""
    out = [0.0] * len(ROLE_LABELS)
    if label in ROLE_LABELS:
        out[ROLE_LABELS.index(label)] = 1.0
    return out


def _option_type_onehot(t: str) -> list[float]:
    out = [0.0] * len(OPTION_TYPES)
    if t in OPTION_TYPES:
        out[OPTION_TYPES.index(t)] = 1.0
    return out


def _intent_onehot(side: str | None) -> list[float]:
    out = [0.0] * len(INTENT_SIDES)
    if side and side in INTENT_SIDES:
        out[INTENT_SIDES.index(side)] = 1.0
    else:
        # Kein Intent / center = center
        out[INTENT_SIDES.index("center")] = 1.0
    return out


def _encode_player(p: dict[str, Any]) -> list[float]:
    pos = p["position"]
    stats = p.get("stats", {})
    return [
        pos[0] / 100.0,
        pos[1] / 100.0,
        *_role_onehot(p.get("position_label", "")),
        p.get("fitness", 100) / 100.0,
        p.get("confidence", 50) / 100.0,
    ]


def _encode_ghost_player() -> list[float]:
    """Padding für fehlende Spieler (< 11 im Team)."""
    return [0.0] * PLAYER_FEAT_DIM


def _encode_option(opt: dict[str, Any]) -> list[float]:
    target = opt.get("target", [0.0, 0.0])
    receiver = opt.get("receiver_id")
    # Receiver-Position ist im Record nicht separat gespeichert, aber wir
    # nutzen target als Proxy (bei Pässen ist target ≈ receiver position).
    has_receiver = 1.0 if receiver else 0.0
    return [
        *_option_type_onehot(opt.get("type", "hold")),
        target[0] / 100.0,
        target[1] / 100.0,
        opt.get("success_chance", 0.5),
        opt.get("reward", 0.0),
        has_receiver,
        target[0] / 100.0 if has_receiver else 0.0,
        target[1] / 100.0 if has_receiver else 0.0,
    ]


# ── Haupt-Encoder ───────────────────────────────────────────────

def encode_sample(rec: dict[str, Any], max_options: int = 16) -> dict[str, torch.Tensor]:
    """Wandelt einen JSONL-Record in Tensor-Features + Label um."""
    ball = rec["ball"]["position"]

    # Global: Ball + Carrier + Score/Time + Intent + Teammates + Opponents
    carrier_feats = _encode_player(rec["carrier"])
    score = rec.get("score", {"team1": 0, "team2": 0})
    team = rec.get("team", 1)
    own_score = score["team1"] if team == 1 else score["team2"]
    opp_score = score["team2"] if team == 1 else score["team1"]
    score_diff = (own_score - opp_score) / 10.0  # normalize, typical range -3..+3

    intent = rec.get("intent") or {}
    intent_side = intent.get("attack_side")
    intent_turns = (intent.get("turns_valid") or 0) / 5.0  # normalize

    # Teammates & Opponents auffüllen
    teammates = rec.get("teammates", [])
    opponents = rec.get("opponents", [])

    team_feats: list[float] = []
    for i in range(10):  # 10 Mitspieler (ohne carrier)
        if i < len(teammates):
            team_feats.extend(_encode_player(teammates[i]))
        else:
            team_feats.extend(_encode_ghost_player())

    opp_feats: list[float] = []
    for i in range(11):
        if i < len(opponents):
            opp_feats.extend(_encode_player(opponents[i]))
        else:
            opp_feats.extend(_encode_ghost_player())

    global_vec = [
        ball[0] / 100.0,
        ball[1] / 100.0,
        1.0,  # possession indicator (wir haben den Ball)
        *carrier_feats,
        score_diff,
        rec.get("game_time_min", 0.0) / 90.0,
        intent_turns,
        float(team),  # wird durch Symmetrie-Augment später spiegelbar
        *_intent_onehot(intent_side),
        *team_feats,
        *opp_feats,
    ]
    assert len(global_vec) == GLOBAL_FEATURE_DIM, \
        f"global dim mismatch: expected {GLOBAL_FEATURE_DIM}, got {len(global_vec)}"

    # Options
    raw_opts = rec.get("options", [])
    num_opts = min(len(raw_opts), max_options)

    option_matrix = torch.zeros(max_options, OPTION_FEATURE_DIM, dtype=torch.float32)
    mask = torch.zeros(max_options, dtype=torch.float32)

    for i in range(num_opts):
        option_matrix[i] = torch.tensor(_encode_option(raw_opts[i]), dtype=torch.float32)
        mask[i] = 1.0

    chosen = rec.get("chosen_option_index", 0)
    if chosen >= max_options:
        chosen = 0  # fallback

    return {
        "global": torch.tensor(global_vec, dtype=torch.float32),
        "options": option_matrix,
        "mask": mask,
        "chosen": chosen,
    }
